CRM/Contact/BAO/GroupContactCache.php

```diff
@@ -372,44 +371,49 @@ public static function remove($groupID = NULL, $onceOnly = TRUE) {
 AND    refresh_date IS NULL
 ";
       }
-    }
-    elseif (is_array($groupID)) {
-      $groupIDs = implode(', ', $groupID);
-      $query = "
-DELETE     g
-FROM       civicrm_group_contact_cache g
-WHERE      g.group_id IN ( $groupIDs )
-";
-      $update = "
-UPDATE civicrm_group g
-SET    cache_date = null,
-       refresh_date = null
-WHERE  id IN ( $groupIDs )
-";
+
+      CRM_Core_DAO::executeQuery($query, $params);
+      if ($refresh) {
+        CRM_Core_DAO::executeQuery($refresh, $params);
+      }
+      // also update the cache_date for these groups
+      CRM_Core_DAO::executeQuery($update, $params);
     }
     else {
-      $query = "
-DELETE     g
-FROM       civicrm_group_contact_cache g
-WHERE      g.group_id = %1
-";
-      $update = "
-UPDATE civicrm_group g
-SET    cache_date = null,
-       refresh_date = null
-WHERE  id = %1
-";
-      $params = array(1 => array($groupID, 'Integer'));
+      foreach ((array) $groupIDs as $groupID) {
+        self::clearGroupContactCache($groupID);
+      }
     }
+  }
 
-    CRM_Core_DAO::executeQuery($query, $params);
+  /**
+   * Function to clear group contact cache and reset the corresponding
+   *  group's cache and refresh date
+   *
+   * @param int $groupID
+   *
+   */
+  public static function clearGroupContactCache($groupID) {
+    $transaction = new CRM_Core_Transaction();
+    $query = "
+    DELETE  g
+      FROM  civicrm_group_contact_cache g
+      WHERE  g.group_id = %1 ";
 
-    if ($refresh) {
-      CRM_Core_DAO::executeQuery($refresh, $params);
-    }
+    $update = "
+  UPDATE civicrm_group g
+    SET    cache_date = null, refresh_date = null
+    WHERE  id = %1 ";
+
+    $params = array(
+      1 => array($groupID, 'Integer'),
+    );
 
+    CRM_Core_DAO::executeQuery($query, $params);
     // also update the cache_date for these groups
     CRM_Core_DAO::executeQuery($update, $params);
+
+    $transaction->commit();
```

Comment:
@totten what impact, if any, does this commit have, given the whole nesting magic?

Reply:
@eileenmcnaughton @totten yes this will arise mysql lock situation. In order to bring further improvement in resetting cache code, here's my patch https://gist.github.com/monishdeb/9cc6eb1b5f1b222c81b52ca9de12201f that also include a change where I replaced ```$transaction->commit()``` with ```$lock->acquire(...)```. Not sure if there's any side effect of lock-nesting, as per the patch:
```
$lock1 = Civi::lockManager()->acquire('data.core.group.refresh'); // at the start of group-cache reset
...
  $lock2 = Civi::lockManager()->acquire("data.core.group.{$groupID}"); // before a group's contact cache is cleared 
...
$lock2->release();
$lock1->release();
```

Reply:
(the patch above contain partial changes on smart-group-rebuild-cache refactoring)